### PR TITLE
Improve compatibility of `upper/lowerUTF8` with Spark

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Functions `upperUTF8` and `lowerUTF8` were previously only able to uppercase / lowercase Cyrillic characters. This limitation is now removed and characters in arbitrary languages are uppercased/lowercased. Example: `SELECT upperUTF8('Süden')` now returns `SÜDEN`.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 44add3d40419028dc286badadf05cfb463551d50 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65761/44add3d40419028dc286badadf05cfb463551d50/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Functions/LowerUpperUTF8Impl.h
Comment (rschu1ze) [inline src/Functions/LowerUpperUTF8Impl.h:64]: This potentially calls `.resize()` on the result structure once per loop iteration (l. 48). What we should do instead is `.reserve()` 2x the current space if the string-to-copy does not fit into the current capacity (i.e. exponential resizing). That way, resizing happens less frequently.

File: src/Functions/LowerUpperUTF8Impl.h
Comment (rschu1ze) [inline src/Functions/LowerUpperUTF8Impl.h:5]: Many developers do "slim" builds, e.g.

```bash
cmake -DENABLE_EMBEDDED_COMPILER=0 -DENABLE_LIBRARIES=0 -DENABLE_LIBURING=1 -DCMAKE_BUILD_TYPE=Debug -S . -B build
```

This build now broke as ifdefs for `USE_ICU` are missing  :-) Like to fix it?

File: tests/queries/0_stateless/00170_lower_upper_utf8.sql
Comment (rschu1ze) [inline tests/queries/0_stateless/00170_lower_upper_utf8.sql:31]: I think you need to update the .reference file for this test, no?

File: src/Functions/LowerUpperUTF8Impl.h
Comment (rschu1ze) [inline src/Functions/LowerUpperUTF8Impl.h:59]: To save heap allocations, we should pull `output` out of the loop and reuse it, i.e. call `.clear()` before each use.

File: tests/queries/0_stateless/00170_lower_upper_utf8.sql
Comment (rschu1ze) [reply tests/queries/0_stateless/00170_lower_upper_utf8.sql:31]: It would be interesting how the ICU-only processing affects performance. If ICU is really really slow, we might want to retain the current implementation under a setting (which is disabled by default).

So, perhaps you like to generate a test table with ASCII-only, with cyrillic-only, and with arbitrary-UTF8-strings and measure performance for the old and new implementations. Thanks!

File: src/Functions/LowerUpperUTF8Impl.h
Comment (taiyang-li) [reply src/Functions/LowerUpperUTF8Impl.h:61]: I found sometimes the output has extra '\0's by  the end. So we need to get content before the first '\0'. We don't need to worry about the issue of remining output because it won't happen for a valid UTF8 string. It we met an invalid UTF8 string, the behavior is not defined and current implementation is also reasonable.

File: src/Functions/LowerUpperUTF8Impl.h
Comment (taiyang-li) [reply src/Functions/LowerUpperUTF8Impl.h:53]: No. from the source code I found a converter will be created by input "UTF-8". But I'm not sure if the converter could be resued between loops.

File: src/Functions/LowerUpperUTF8Impl.h
Comment (taiyang-li) [reply src/Functions/LowerUpperUTF8Impl.h:64]: resize has 2 steps 
1. reserve 2x current space if current space is not enough
2. execute `resize_assume_reserved` 

The second step is too trivial. The first step doesn't happen frequently. So I believe it it not necessary to `reserve` in advance because you never known what is the exact size of utf8 string after upper/lower conversion.

File: tests/queries/0_stateless/00170_lower_upper_utf8.sql
Comment (taiyang-li) [reply tests/queries/0_stateless/00170_lower_upper_utf8.sql:33]: No significant performance degression was found 
```
select upperUTF8('ır ır ır ır ır ır ır ır ır ır ır ır ır ır ır ') from numbers(1000000000) format Null;

New:
0 rows in set. Elapsed: 0.405 sec. Processed 1.00 billion rows, 8.00 GB (2.47 billion rows/s., 19.73 GB/s.)
0 rows in set. Elapsed: 0.405 sec. Processed 1.00 billion rows, 8.00 GB (2.47 billion rows/s., 19.73 GB/s.)
0 rows in set. Elapsed: 0.410 sec. Processed 1.00 billion rows, 8.00 GB (2.44 billion rows/s., 19.49 GB/s.)

Old
0 rows in set. Elapsed: 0.383 sec. Processed 1.00 billion rows, 8.00 GB (2.61 billion rows/s., 20.87 GB/s.)
0 rows in set. Elapsed: 0.403 sec. Processed 1.00 billion rows, 8.00 GB (2.48 billion rows/s., 19.83 GB/s.)
0 rows in set. Elapsed: 0.386 sec. Processed 1.00 billion rows, 8.00 GB (2.59 billion rows/s., 20.73 GB/s.)
```

Comment (taiyang-li) [reply tests/queries/0_stateless/00170_lower_upper_utf8.sql:33]:
``` sql 
select upperUTF8('skshpfdhpwoehowehoehfopehopehop') from numbers(1000000000) format Null;
New:
0 rows in set. Elapsed: 0.408 sec. Processed 1.00 billion rows, 8.00 GB (2.45 billion rows/s., 19.63 GB/s.)
0 rows in set. Elapsed: 0.383 sec. Processed 1.00 billion rows, 8.00 GB (2.61 billion rows/s., 20.87 GB/s.)
0 rows in set. Elapsed: 0.387 sec. Processed 1.00 billion rows, 8.00 GB (2.58 billion rows/s., 20.65 GB/s.)
Old: 
0 rows in set. Elapsed: 0.384 sec. Processed 1.00 billion rows, 8.00 GB (2.60 billion rows/s., 20.83 GB/s.)
0 rows in set. Elapsed: 0.390 sec. Processed 1.00 billion rows, 8.00 GB (2.57 billion rows/s., 20.53 GB/s.)
0 rows in set. Elapsed: 0.389 sec. Processed 1.00 billion rows, 8.00 GB (2.57 billion rows/s., 20.56 GB/s.)
```

File: src/Functions/LowerUpperUTF8Impl.h
Comment (rschu1ze) [reply src/Functions/LowerUpperUTF8Impl.h:60]: Ah, I was not aware that `resize()` does an exponential resize internally. We are good then.

**Comment by rschu1ze** [comment]
@taiyang-li Thanks, the PR looks good to me.

While reviewing, I made some small cleanups but since I get a 403 (permission) error when I push to your repository, I pushed to my fork: https://github.com/rschu1ze/ClickHouse/commits/improve_upper_lower/
It would be great if you could pick the top commit into this PR.
(most notably, it removes the indentation for includes which I originally proposed, it looks better without indentation)

**Comment by taiyang-li** [comment]
> @taiyang-li Thanks, the PR looks good to me.
> 
> While reviewing, I made some small cleanups but since I get a 403 (permission) error when I push to your repository, I pushed to my fork: https://github.com/rschu1ze/ClickHouse/commits/improve_upper_lower/ It would be great if you could pick the top commit into this PR. (most notably, it removes the indentation for includes which I originally proposed, it looks better without indentation)

Thanks! It is done.

**Comment by rschu1ze** [comment]
@taiyang-li Some test failures (stateless tests) look related.

**Comment by rschu1ze** [comment]
@taiyang-li About tests:
- Test "02415_all_new_functions_must_be_documented": I think this will be fixed by adding in-source documentation, see here for an example how to do that: https://github.com/ClickHouse/ClickHouse/blob/master/src/Functions/generateSnowflakeID.cpp#L208
- There are asan issues found [here](https://s3.amazonaws.com/clickhouse-test-reports/65761/a143b4b7ff40b8e81240fa8c1c7e61ee4f5984c3/stateless_tests_flaky_check__asan_.html) and [here](https://s3.amazonaws.com/clickhouse-test-reports/65761/a143b4b7ff40b8e81240fa8c1c7e61ee4f5984c3/stateless_tests_flaky_check__asan_/stderr.log) :-(

To reproduce the latter, run `cmake -DENABLE_EMBEDDED_COMPILER=0 -DENABLE_LIBRARIES=0 -DENABLE_LIBURING=1  -DCMAKE_BUILD_TYPE=None -DSANITIZE=address -S . -B build` and then make a build.

**Comment by rschu1ze** [comment]
@taiyang-li Unfortunately, there is another merge conflict.

**Comment by rschu1ze** [comment]
Some more issues :-(

https://s3.amazonaws.com/clickhouse-test-reports/65761/8de50d2ca6c2778a93500cf7eb72ad327ad57ec1/stateless_tests_flaky_check__asan_.html

**Comment by rschu1ze** [comment]
@taiyang-li There seem to be some msan issues :-(

**Comment by rschu1ze** [comment]
@taiyang-li Query `SELECT lowerUTF8('aaaaАБВГAAAAaaAA') = 'aaaaабвгaaaaaaaa'` produces this: `WARNING: MemorySanitizer: use-of-uninitialized-value`: https://s3.amazonaws.com/clickhouse-test-reports/65761/a273def46e478e465af2078a72dd49ae3a685fb5/ast_fuzzer__msan_.html

To reproduce, build an msan build: `cmake -DENABLE_EMBEDDED_COMPILER=0 -DENABLE_LIBRARIES=1  -DCMAKE_BUILD_TYPE=None -DSANITIZE=memory -S . -B build_msan'`

**Comment by taiyang-li** [comment]
@rschu1ze I don't quite understand why `s` is uninitialized 

https://github.com/unicode-org/icu/blob/a56dde820dc35665a66f2e9ee8ba58e75049b668/icu4c/source/common/ustrcase.cpp#L269-L282

**Comment by rschu1ze** [comment]
I unfortunately don't have the time today to check in detail, if needed I can perhaps check on the weekend.

**Comment by taiyang-li** [comment]
> I unfortunately don't have the time today to check in detail, if needed I can perhaps check on the weekend.

Thank you very much. I'll look into it first.

**Comment by taiyang-li** [comment]
@rschu1ze it was fixed in https://github.com/unicode-org/icu/pull/3061

**Comment by rschu1ze** [comment]
@taiyang-li Thanks for the fix. I added a comment to https://github.com/unicode-org/icu/pull/3061 . Totally not to discourage you, but let's first be sure that the upstream patch to ICU 1. really fixes the issue (based on looking at the test results in this PR here) and 2. the fix is merged in upstream ICU (asking because there were situations in the past (other PRs) where the ClickHouse PR was prematurely merged but the upstream project requested changes to an associated fix, so the final fix had to be ported to ClickHouse to keep things in-sync).

File: docs/en/sql-reference/functions/string-functions.md
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/string-functions.md:571]: No. This still holds.

File: docs/en/sql-reference/functions/string-functions.md
Comment (alexey-milovidov) [inline docs/en/sql-reference/functions/string-functions.md:609]: No. This still holds.

File: src/Functions/LowerUpperUTF8Impl.h
Comment (alexey-milovidov) [inline src/Functions/LowerUpperUTF8Impl.h:37]: A fast-path for Latin+Cyrillic should remain. Add a performance test.

**Review by alexey-milovidov — DISMISSED** [review]
.

File: src/Functions/LowerUpperUTF8Impl.h
Comment (taiyang-li) [reply src/Functions/LowerUpperUTF8Impl.h:37]: It is tested manully in https://github.com/ClickHouse/ClickHouse/pull/65761#discussion_r1660435973 and https://github.com/ClickHouse/ClickHouse/pull/65761#discussion_r1660438004     The new implementation has little impact on perforamance.

**Comment by taiyang-li** [comment]
@rschu1ze  the tests seem all right. Should we wait for the fix of icu lib to be merge first before merging this PR ?

**Comment by rschu1ze** [comment]
@taiyang-li Yes, we should wait until both
- https://github.com/ClickHouse/ClickHouse/pull/66474 and
- https://github.com/unicode-org/icu/pull/3061

are merged. The former has right now less importance than fixing ClickHouse's CI tests which are currently not in a good shape. The situation will hopefully improve in 1-2 weeks.

**Comment by rschu1ze** [comment]
@taiyang-li https://github.com/unicode-org/icu/pull/3061 was approved. But we don't need to wait until a new ICU release. I forked tag `release-75-1` as branch `ClickHouse/release-71-1` and picked your fix on top: https://github.com/ClickHouse/icu/tree/ClickHouse/release-75-1

Could you please remove
- merge from `master`, then
- remove the current ICU submodule, then
- add ClickHouse ICU fork repository as submodule, and
- switch the submodule to `ClickHouse/release-75-1`?

Thanks.

**Review by rschu1ze — APPROVED** [review]

**Comment by taiyang-li** [comment]
@rschu1ze May I ask why reverted this pr after merged it ?

**Comment by rschu1ze** [comment]
As you might know, ClickHouse has a public repository (this one) and a private repository. PRs in the public repository can only be merged if the same PR does not create merge conflicts in the private repository. This was not the case for this PR, however I hit the merge button, and only noticed the conflict then, and reverted this PR (as per development policy).

(The conflict was actually minor: the replaced ICU submodule appeared at the end of .gitmodules in this PR, but the private repository had another submodule at the end already.)

Nevermind, I re-pushed your PR here and will take care of merging it:

**Pull request merged**
